File: android/app/src/main/java/ru/dygdyg/trackanime/ProxyFallback.java
package ru.dygdyg.trackanime;

import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.webkit.ProxyConfig;
import androidx.webkit.ProxyController;
import androidx.webkit.WebViewFeature;

/** Applies an optional HTTP proxy to the WebView only after direct mirrors fail. */
final class ProxyFallback {
    enum Mode { NONE, SERVER, MANUAL }

    static final class Settings {
        final Mode mode;
        final String host;
        final int port;
        final String username;
        final String password;

        Settings(Mode mode, String host, int port, String username, String password) {
            this.mode = mode;
            this.host = host;
            this.port = port;
            this.username = username;
            this.password = password;
        }
    }

    /** Result of parsing a {@code taproxy://} deep link. */
    static final class DeepLinkResult {
        final Settings settings;
        final String error;

        DeepLinkResult(Settings settings, String error) {
            this.settings = settings;
            this.error = error;
        }

        boolean isOk() {
            return settings != null && error == null;
        }
    }

    private static final String LOG_TAG = "TrackAnimeProxy";
    private static final String MODE_KEY = "proxy-mode";
    private static final String HOST_KEY = "proxy-manual-host";
    private static final String PORT_KEY = "proxy-manual-port";
    private static final String USERNAME_KEY = "proxy-manual-username";
    private static final String PASSWORD_KEY = "proxy-manual-password";

    private final Handler handler = new Handler(Looper.getMainLooper());
    private final SharedPreferences preferences;
    private boolean enabled;
    private Settings enabledSettings;

    ProxyFallback(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    /**
     * Supported forms:
     * <ul>
     *   <li>{@code taproxy://host:port}</li>
     *   <li>{@code taproxy://user:pass@host:port}</li>
     *   <li>{@code taproxy://host:port?user=...&pass=...} ({@code username}/{@code password} also accepted)</li>
     *   <li>{@code taproxy://clear} / {@code none} / {@code off} — disable proxy</li>
     *   <li>{@code taproxy://server} — use built-in server proxy from local.properties</li>
     * </ul>
     */
    static DeepLinkResult parseDeepLink(Uri uri) {
        if (uri == null || !"taproxy".equalsIgnoreCase(uri.getScheme())) {
            return new DeepLinkResult(null, "Неверная ссылка прокси.");
        }

        String host = uri.getHost();
        if (host == null || host.isEmpty()) {
            return new DeepLinkResult(null, "Укажите адрес прокси в ссылке.");
        }

        String hostLower = host.toLowerCase();
        if ("clear".equals(hostLower) || "none".equals(hostLower) || "off".equals(hostLower)
                || "disable".equals(hostLower)) {
            return new DeepLinkResult(new Settings(Mode.NONE, "", 0, "", ""), null);
        }
        if ("server".equals(hostLower) || "default".equals(hostLower)) {
            return new DeepLinkResult(new Settings(Mode.SERVER, "", 0, "", ""), null);
        }

        int port = uri.getPort();
        if (port < 0) {
            String portQuery = firstQuery(uri, "port");
            if (portQuery != null) {
                try {
                    port = Integer.parseInt(portQuery.trim());
                } catch (NumberFormatException ignored) {
                    port = -1;
                }
            }
        }
        if (port < 1 || port > 65535) {
            return new DeepLinkResult(null, "Укажите порт прокси от 1 до 65535.");
        }

        String username = "";
        String password = "";
        String userInfo = uri.getUserInfo();
        if (userInfo != null && !userInfo.isEmpty()) {
            int separator = userInfo.indexOf(':');
            if (separator < 0) {
                username = Uri.decode(userInfo);
            } else {
                username = Uri.decode(userInfo.substring(0, separator));
                password = Uri.decode(userInfo.substring(separator + 1));
            }
        }
        String queryUser = firstQuery(uri, "user", "username");
        String queryPass = firstQuery(uri, "pass", "password");
        if (queryUser != null) username = queryUser;
        if (queryPass != null) password = queryPass;

        return new DeepLinkResult(
                new Settings(Mode.MANUAL, host.trim(), port, username == null ? "" : username,
                        password == null ? "" : password),
                null
        );
    }

    /** Builds a shareable {@code taproxy://} link for the given manual proxy fields. */
    static String buildDeepLink(String host, int port, String username, String password) {
        String trimmedHost = host == null ? "" : host.trim();
        if (trimmedHost.isEmpty() || port < 1 || port > 65535) return null;
        String user = username == null ? "" : username.trim();
        String pass = password == null ? "" : password;
        Uri.Builder builder = new Uri.Builder().scheme("taproxy");
        if (!user.isEmpty() || !pass.isEmpty()) {
            builder.encodedAuthority(
                    Uri.encode(user) + ":" + Uri.encode(pass) + "@" + trimmedHost + ":" + port);
        } else {
            builder.encodedAuthority(trimmedHost + ":" + port);
        }
        return builder.build().toString();
    }

    private static String firstQuery(Uri uri, String... keys) {
        for (String key : keys) {
            String value = uri.getQueryParameter(key);
            if (value != null) return value;
        }
        return null;
    }

    boolean isConfigured() {
        Settings settings = getSettings();
        return settings.mode != Mode.NONE && !settings.host.isEmpty() && settings.port > 0 && settings.port <= 65535;
    }

    Settings getSettings() {
        Mode mode;
        try {
            mode = Mode.valueOf(preferences.getString(MODE_KEY, Mode.SERVER.name()));
        } catch (IllegalArgumentException | NullPointerException ignored) {
            mode = Mode.SERVER;
        }

        if (mode == Mode.MANUAL) {
            return new Settings(mode, preferences.getString(HOST_KEY, "").trim(), preferences.getInt(PORT_KEY, 0),
                    preferences.getString(USERNAME_KEY, ""), preferences.getString(PASSWORD_KEY, ""));
        }
        if (mode == Mode.SERVER) {
            return new Settings(mode, BuildConfig.FALLBACK_PROXY_HOST, BuildConfig.FALLBACK_PROXY_PORT,
                    BuildConfig.FALLBACK_PROXY_USERNAME, BuildConfig.FALLBACK_PROXY_PASSWORD);
        }
        return new Settings(mode, "", 0, "", "");
    }

    void saveSettings(Settings settings) {
        preferences.edit().putString(MODE_KEY, settings.mode.name()).putString(HOST_KEY, settings.host.trim())
                .putInt(PORT_KEY, settings.port).putString(USERNAME_KEY, settings.username)
                .putString(PASSWORD_KEY, settings.password).apply();
    }

    boolean matchesEnabledProxy(String host) {
        return enabled && enabledSettings != null && enabledSettings.host.equalsIgnoreCase(host);
    }

    String getEnabledUsername() { return enabledSettings == null ? "" : enabledSettings.username; }

    String getEnabledPassword() { return enabledSettings == null ? "" : enabledSettings.password; }

    void clearOverride(Runnable onReady) {
        enabled = false;
        enabledSettings = null;
        if (!WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE)) {
            onReady.run();
            return;
        }
        try {
            ProxyController.getInstance().clearProxyOverride(handler::post, () -> handler.post(onReady));
        } catch (IllegalArgumentException | UnsupportedOperationException error) {
            Log.w(LOG_TAG, "Could not clear proxy override.", error);
            onReady.run();
        }
    }

    void enable(Runnable onReady) {
        Settings settings = getSettings();
        if (enabled && enabledSettings != null && enabledSettings.host.equals(settings.host)
                && enabledSettings.port == settings.port && enabledSettings.username.equals(settings.username)
                && enabledSettings.password.equals(settings.password)) {
            onReady.run();
            return;
        }
        if (!isConfigured() || !WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE)) {
            Log.w(LOG_TAG, "Proxy fallback is unavailable on this device or is disabled.");
            onReady.run();
            return;
        }

        ProxyConfig config = new ProxyConfig.Builder().addProxyRule(
                settings.host + ":" + settings.port, ProxyConfig.MATCH_ALL_SCHEMES).build();
        try {
            ProxyController.getInstance().setProxyOverride(config, handler::post, () -> {
                enabled = true;
                enabledSettings = settings;
                Log.i(LOG_TAG, "Proxy fallback enabled: " + settings.mode);
                handler.post(onReady);
            });
        } catch (IllegalArgumentException | UnsupportedOperationException error) {
            Log.w(LOG_TAG, "Could not enable proxy fallback.", error);
            onReady.run();
        }
    }
}

File: windows/TrackAnime/ProxyFallback.cs
using System.IO;
using System.Text;

namespace TrackAnime;

public enum ProxyMode
{
    None,
    Server,
    Manual,
}

public sealed record ProxySettings(ProxyMode Mode, string Host, int Port, string Username, string Password);

internal sealed class ProxyDeepLinkResult
{
    public ProxySettings? Settings { get; init; }
    public string? Error { get; init; }
    public bool IsOk => Settings is not null && Error is null;
}

internal static class ProxyFallback
{
    public static ProxySettings Resolve(AppPreferences preferences, LocalProperties local)
    {
        return preferences.ProxyMode switch
        {
            ProxyMode.Manual => new ProxySettings(
                ProxyMode.Manual,
                preferences.ProxyHost.Trim(),
                preferences.ProxyPort,
                preferences.ProxyUsername,
                preferences.ProxyPassword),
            ProxyMode.Server => new ProxySettings(
                ProxyMode.Server,
                local.ProxyHost,
                local.ProxyPort,
                local.ProxyUsername,
                local.ProxyPassword),
            _ => new ProxySettings(ProxyMode.None, "", 0, "", ""),
        };
    }

    public static bool IsConfigured(ProxySettings settings) =>
        settings.Mode != ProxyMode.None
        && !string.IsNullOrWhiteSpace(settings.Host)
        && settings.Port is > 0 and <= 65535;

    public static string? BuildBrowserArguments(ProxySettings settings)
    {
        if (!IsConfigured(settings)) return null;
        return $"--proxy-server=http://{settings.Host}:{settings.Port}";
    }

    public static ProxyDeepLinkResult ParseDeepLink(Uri uri)
    {
        if (!string.Equals(uri.Scheme, "taproxy", StringComparison.OrdinalIgnoreCase))
        {
            return new ProxyDeepLinkResult { Error = "Неверная ссылка прокси." };
        }

        var host = uri.Host;
        if (string.IsNullOrWhiteSpace(host))
        {
            return new ProxyDeepLinkResult { Error = "Укажите адрес прокси в ссылке." };
        }

        var hostLower = host.ToLowerInvariant();
        if (hostLower is "clear" or "none" or "off" or "disable")
        {
            return new ProxyDeepLinkResult { Settings = new ProxySettings(ProxyMode.None, "", 0, "", "") };
        }
        if (hostLower is "server" or "default")
        {
            return new ProxyDeepLinkResult { Settings = new ProxySettings(ProxyMode.Server, "", 0, "", "") };
        }

        var port = uri.Port;
        if (port < 0)
        {
            var portQuery = FirstQuery(uri, "port");
            if (portQuery is not null && int.TryParse(portQuery.Trim(), out var parsed))
            {
                port = parsed;
            }
        }
        if (port is < 1 or > 65535)
        {
            return new ProxyDeepLinkResult { Error = "Укажите порт прокси от 1 до 65535." };
        }

        var username = "";
        var password = "";
        if (!string.IsNullOrEmpty(uri.UserInfo))
        {
            var separator = uri.UserInfo.IndexOf(':');
            if (separator < 0)
            {
                username = Uri.UnescapeDataString(uri.UserInfo);
            }
            else
            {
                username = Uri.UnescapeDataString(uri.UserInfo[..separator]);
                password = Uri.UnescapeDataString(uri.UserInfo[(separator + 1)..]);
            }
        }
        var queryUser = FirstQuery(uri, "user", "username");
        var queryPass = FirstQuery(uri, "pass", "password");
        if (queryUser is not null) username = queryUser;
        if (queryPass is not null) password = queryPass;

        return new ProxyDeepLinkResult
        {
            Settings = new ProxySettings(ProxyMode.Manual, host.Trim(), port, username, password),
        };
    }

    public static string? BuildDeepLink(string host, int port, string username, string password)
    {
        var trimmedHost = host.Trim();
        if (string.IsNullOrEmpty(trimmedHost) || port is < 1 or > 65535) return null;
        var user = username.Trim();
        var pass = password ?? "";
        if (!string.IsNullOrEmpty(user) || !string.IsNullOrEmpty(pass))
        {
            return $"taproxy://{Uri.EscapeDataString(user)}:{Uri.EscapeDataString(pass)}@{trimmedHost}:{port}";
        }
        return $"taproxy://{trimmedHost}:{port}";
    }

    private static string? FirstQuery(Uri uri, params string[] keys)
    {
        var query = uri.Query;
        if (string.IsNullOrEmpty(query)) return null;
        var parts = query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries);
        foreach (var key in keys)
        {
            foreach (var part in parts)
            {
                var eq = part.IndexOf('=');
                var name = eq < 0 ? part : part[..eq];
                if (!string.Equals(Uri.UnescapeDataString(name), key, StringComparison.OrdinalIgnoreCase)) continue;
                return eq < 0 ? "" : Uri.UnescapeDataString(part[(eq + 1)..]);
            }
        }
        return null;
    }
}

public sealed class LocalProperties
{
    public string ProxyHost { get; private set; } = "";
    public int ProxyPort { get; private set; }
    public string ProxyUsername { get; private set; } = "";
    public string ProxyPassword { get; private set; } = "";

    public static LocalProperties Load()
    {
        var result = new LocalProperties();
        foreach (var candidate in CandidatePaths())
        {
            if (!File.Exists(candidate)) continue;
            try
            {
                foreach (var rawLine in File.ReadAllLines(candidate, Encoding.UTF8))
                {
                    var line = rawLine.Trim();
                    if (line.Length == 0 || line.StartsWith('#') || line.StartsWith('!')) continue;
                    var eq = line.IndexOf('=');
                    if (eq <= 0) continue;
                    var key = line[..eq].Trim();
                    var value = line[(eq + 1)..].Trim();
                    switch (key)
                    {
                        case "trackAnimeProxyHost":
                            result.ProxyHost = value;
                            break;
                        case "trackAnimeProxyPort" when int.TryParse(value, out var port):
                            result.ProxyPort = port;
                            break;
                        case "trackAnimeProxyUsername":
                            result.ProxyUsername = value;
                            break;
                        case "trackAnimeProxyPassword":
                            result.ProxyPassword = value;
                            break;
                    }
                }
                break;
            }
            catch
            {
                // Ignore unreadable local.properties.
            }
        }
        return result;
    }

    private static IEnumerable<string> CandidatePaths()
    {
        var baseDir = AppContext.BaseDirectory;
        yield return Path.Combine(baseDir, "local.properties");
        yield return Path.GetFullPath(Path.Combine(baseDir, "..", "..", "..", "local.properties"));
        yield return Path.GetFullPath(Path.Combine(baseDir, "..", "..", "..", "..", "local.properties"));
        yield return Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "local.properties");
    }
}
